lib/GroupManager.php

```diff
@@ -0,0 +1,258 @@
+		$this->settings = $settings;
+	}
+
+	private function getGroupsToRemove(array $samlGroups, array $assignedGroups): array {
+		$groupsToRemove = [];
+		foreach ($assignedGroups as $group) {
+			// if group is not supplied by SAML and group has SAML backend
+			if (!in_array($group->getGID(), $samlGroups) && $this->hasSamlBackend($group)) {
+				$groupsToRemove[] = $group->getGID();
+			}
+		}
+		return $groupsToRemove;
+	}
+
+	private function getGroupsToAdd(array $samlGroups, array $assignedGroupIds): array {
+		$groupsToAdd = [];
+		foreach ($samlGroups as $group) {
+			// if user is not assigend to the group or the provided group has a non SAML backend
+			if (!in_array($group, $assignedGroupIds) || !$this->hasSamlBackend($this->groupManager->get($group))) {
+				$groupsToAdd[] = $group;
+			}
+		}
+		return $groupsToAdd;
+	}
+
+	public function replaceGroups(string $uid, array $samlGroups): void {
+		$user = $this->userManager->get($uid);
+		if ($user === null) {
+			return;
+		}
+		$this->translateGroupToIds($samlGroups);
+		$assignedGroups = $this->groupManager->getUserGroups($user);
+		$assignedGroupIds = array_map(function (IGroup $group) {
+			return $group->getGID();
+		}, $assignedGroups);
+		$groupsToRemove = $this->getGroupsToRemove($samlGroups, $assignedGroups);
+		$groupsToAdd = $this->getGroupsToAdd($samlGroups, $assignedGroupIds);
+		$this->removeGroups($user, $groupsToRemove);
+		$this->addGroups($user, $groupsToAdd);
+	}
+
+	protected function translateGroupToIds(array &$samlGroups): void {
+		array_walk($samlGroups, function (&$gid) {
+			$altGid = $this->ownGroupBackend->groupExistsWithDifferentGid($gid);
+			if ($altGid !== null) {
+				$gid = $altGid;
+			}
+		});
+	}
+
+	public function removeGroups(IUser $user, array $groupIds): void {
+		foreach ($groupIds as $gid) {
+			$this->removeGroup($user, $gid);
+		}
+	}
+
+	public function removeGroup(IUser $user, string $gid): void {
+		$group = $this->groupManager->get($gid);
+		if ($group === null) {
+			return;
+		}
+		$this->ownGroupBackend->removeFromGroup($user->getUID(), $group->getGID());
+		if ($this->ownGroupBackend->countUsersInGroup($gid) === 0) {
+			$this->ownGroupBackend->deleteGroup($group->getGID());
+		}
+	}
+
+	public function addGroups(IUser $user, $groupIds): void {
+		foreach ($groupIds as $gid) {
+			$this->addGroup($user, $gid);
+		}
+	}
+
+	public function addGroup(IUser $user, string $gid): void {
+		try {
+			$group = $this->findGroup($gid);
+		} catch (\RuntimeException $e) {
+			if ($e->getCode() === 1) {
+				$group = $this->createGroupInBackend($gid);
+			} elseif ($e->getCode() === 2) {
+				//FIXME: probably need config flag. Previous to 17, gid was used as displayname
+				$providerId = $this->settings->getProviderId();
+				$settings = $this->settings->get($providerId);
+				$groupPrefix = isset($settings['saml-attribute-mapping-group_mapping_prefix'])
+					? $settings['saml-attribute-mapping-group_mapping_prefix'] : 'SAML_';
```

Comment:
for the configuration itself is of course IdP configuration dependent, it might be worthwhile to include the IdP-Id in the prefix. Otherwise it might collide with other SAML groups… unless this behaviour is desired (I doubt)?

Reply:
On the other handsight, it might become tricky on migration scenarios. I leave it as is.